### ISSUE TITLE
Reading images concurrently in a threaded setup may result in a segfault

### ISSUE THREAD

**Issue by aomader** [SimpleITK/SimpleITK#74]
Not quite sure why this is the case, but reading multiple jpeg files concurrently in a threaded setup may result in a segfault. Compare the stacktrace below.

```
*** Error in `/root/oli/rfl/env/bin/python3': double free or corruption (fasttop): 0x00007fc1046223e0 ***
======= Backtrace: =========
/usr/lib/libc.so.6(+0x70c4b)[0x7fc1f72fec4b]
/usr/lib/libc.so.6(+0x76fe6)[0x7fc1f7304fe6]
/usr/lib/libc.so.6(+0x777de)[0x7fc1f73057de]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(H5MM_xfree+0xe)[0x7fc174c86ece]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(+0x9470917)[0x7fc174b09917]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(H5E_clear_stack+0x35)[0x7fc174b09a85]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(H5Fis_hdf5+0x2b)[0x7fc174b0c6bb]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(_ZN2H56H5File6isHdf5EPKc+0xd)[0x7fc174aca0fd]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(_ZN3itk11HDF5ImageIO11CanReadFileEPKc+0x110)[0x7fc1747a4ba0]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(_ZN3itk14ImageIOFactory13CreateImageIOEPKcNS0_12FileModeTypeE+0x25c)[0x7fc174f4765c]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(+0x3865515)[0x7fc16eefe515]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(+0x370c569)[0x7fc16eda5569]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(+0x370cc28)[0x7fc16eda5c28]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(+0x335c038)[0x7fc16e9f5038]
/root/oli/rfl/env/lib/python3.5/site-packages/SimpleITK/_SimpleITK.so(+0x335cf35)[0x7fc16e9f5f35]
/usr/lib/libpython3.5m.so.1.0(PyCFunction_Call+0xc9)[0x7fc1f78f8bc9]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalFrameEx+0x799f)[0x7fc1f796fa3f]
/usr/lib/libpython3.5m.so.1.0(+0x1285e3)[0x7fc1f79715e3]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalFrameEx+0x5d8e)[0x7fc1f796de2e]
/usr/lib/libpython3.5m.so.1.0(+0x1285e3)[0x7fc1f79715e3]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalCodeEx+0x23)[0x7fc1f79716c3]
/usr/lib/libpython3.5m.so.1.0(+0x92fd8)[0x7fc1f78dbfd8]
/usr/lib/libpython3.5m.so.1.0(PyObject_Call+0x47)[0x7fc1f78b12b7]
/usr/lib/libpython3.5m.so.1.0(+0x7ed14)[0x7fc1f78c7d14]
/usr/lib/libpython3.5m.so.1.0(PyObject_Call+0x47)[0x7fc1f78b12b7]
/usr/lib/libpython3.5m.so.1.0(+0x185ee2)[0x7fc1f79ceee2]
/usr/lib/libpython3.5m.so.1.0(PyObject_Call+0x47)[0x7fc1f78b12b7]
/usr/lib/libpython3.5m.so.1.0(+0x119a1e)[0x7fc1f7962a1e]
/usr/lib/libpython3.5m.so.1.0(+0x95946)[0x7fc1f78de946]
/usr/lib/libpython3.5m.so.1.0(+0x95c00)[0x7fc1f78dec00]
/usr/lib/libpython3.5m.so.1.0(+0xc3475)[0x7fc1f790c475]
/usr/lib/libpython3.5m.so.1.0(PyObject_Call+0x47)[0x7fc1f78b12b7]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalFrameEx+0x3d6b)[0x7fc1f796be0b]
/usr/lib/libpython3.5m.so.1.0(+0x1285e3)[0x7fc1f79715e3]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalCodeEx+0x23)[0x7fc1f79716c3]
/usr/lib/libpython3.5m.so.1.0(+0x930c5)[0x7fc1f78dc0c5]
/usr/lib/libpython3.5m.so.1.0(PyObject_Call+0x47)[0x7fc1f78b12b7]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalFrameEx+0xee6)[0x7fc1f7968f86]
/usr/lib/libpython3.5m.so.1.0(+0x1285e3)[0x7fc1f79715e3]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalCodeEx+0x23)[0x7fc1f79716c3]
/usr/lib/libpython3.5m.so.1.0(+0x930c5)[0x7fc1f78dc0c5]
/usr/lib/libpython3.5m.so.1.0(PyObject_Call+0x47)[0x7fc1f78b12b7]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalFrameEx+0xee6)[0x7fc1f7968f86]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalFrameEx+0x88ca)[0x7fc1f797096a]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalFrameEx+0x88ca)[0x7fc1f797096a]
/usr/lib/libpython3.5m.so.1.0(+0x1285e3)[0x7fc1f79715e3]
/usr/lib/libpython3.5m.so.1.0(PyEval_EvalCodeEx+0x23)[0x7fc1f79716c3]
/usr/lib/libpython3.5m.so.1.0(+0x92fd8)[0x7fc1f78dbfd8]
/usr/lib/libpython3.5m.so.1.0(PyObject_Call+0x47)[0x7fc1f78b12b7]
/usr/lib/libpython3.5m.so.1.0(+0x7ed14)[0x7fc1f78c7d14]
/usr/lib/libpython3.5m.so.1.0(PyObject_Call+0x47)[0x7fc1f78b12b7]
/usr/lib/libpython3.5m.so.1.0(PyEval_CallObjectWithKeywords+0x47)[0x7fc1f7967da7]
/usr/lib/libpython3.5m.so.1.0(+0x1629a2)[0x7fc1f79ab9a2]
/usr/lib/libpthread.so.0(+0x7454)[0x7fc1f7633454]
/usr/lib/libc.so.6(clone+0x5f)[0x7fc1f73767df]
```

**Comment by blowekamp** [comment]
Do you have sample code to reproduce the problem? Is it occasional or does it always happen?

**Comment by blowekamp** [comment]
This call stack stays this is an HD5 issue. The HDF5 library has a configuration option to ensure it's thread safe:
 https://support.hdfgroup.org/hdf5-quest.html#tsafe

I have had success concurrently reading tiff files. There may be other issues with other ImageIO types, caution should be used with concurrent IO. We should ensure that the mechanism to determine the ImageIO type is concurrent thread safe.

**Comment by b52** [comment]
Here is some sample code that reproduces the problem on my machine always, assuming you have a few jpeg images.

- ArchLinux
- Python 3.5.2
- SimpleITK 0.10.0

```python
import glob
import SimpleITK as sitk
from multiprocessing.pool import ThreadPool

files = glob.glob('/home/oli/Projects/rfl/tests/data/2d/images/*.jpg')
files = files * 1000

def load_file(path):
    print('Loading file', path)
    image = sitk.ReadImage(path)
    data = sitk.GetArrayFromImage(image)

with ThreadPool() as p:
    list(p.imap(load_file, files))
```

I also don't understand why HDF5 is involved in reading jpeg images. Additionally, I have the same problem with other image formats (2D as well as 3D).

**Comment by blowekamp** [comment]
The HDF5 ImageIO is used when determining the file format. Here is a first pass to try to address the issue in ITK:
http://review.source.kitware.com/#/c/21811/

**Comment by b52** [comment]
I think you references the wrong ticket @blowekamp, since the linked one just replicates the HDF5 cmake options in ITK. These changes shouldn't be related to this issue.

**Comment by b52** [comment]
Yeah, that looks like a reasonable thing to do. But it probably takes some time until this lands in SimpleITK in form of a new release, I would assume.

Does this patch might fix this [issue](https://github.com/SimpleITK/SimpleITK/issues/45) as well?

**Comment by blowekamp** [comment]
We are writing a test for multi-threading I/O with a couple different formant to determine how stable this operation is and for what formats.

I expect that ITK patch will be in the next branch of SimpleITK today. It's not too hard to build the SimpleITK superbuild from the source. It just takes a little time. When this ITK patch is merged and in SimpleITK's next I'd recommend compiling it and trying the potential fix out for your files.

The next SimpleITK release will probably be at the end of next month. Also there are some nightly binaries on the [dashboard](https://open.cdash.org/index.php?project=Insight) for the master branch, which may be of help.

**Comment by blowekamp** [comment]
I have written a test and experimented with the fix. It seems to resolve the problem. The usability of concurrent reading and writing of image in SimpleITK is subject to the thread-safety of the individual ITK ImageIO, and the underlying third-party library used for IO. So this feature can be used with caution, and testing.

When the underlying C++ ITK code is called from SimpleITK the Python GIL is unlocked. This enables light weight Python threads to be used effectively, unlike with normal pure Python code. While many cases the IO will be bound by storage access, the approach appears to be quite effective.

I have a directory of the Visible Human as 1878 png files.

Here is the performance improvement I see:


In [9]: %timeit -n1 -r3 p.map(sitk.ReadImage, files)
1 loop, best of 3: 2.86 s per loop

In [10]: %timeit -n1 -r1 map(sitk.ReadImage, files)
1 loop, best of 1: 1min 28s per loop

**Comment by nslay** [comment]
This issue is still haunting me in SimpleITK 2.2.0. It doesn't matter if I use sitk.ReadImage, sitk.ImageFileReader or if I manually select an ImageIO. I've also tried newer versions of HDF5 since that was identified in this old issue (though unsure if SimpleITK python package is using it). I am quite sure this is a problem in ITK too.

My workaround is to use nibabel 😢

**Comment by blowekamp** [comment]
We could try rebuilding ITK/SimpleITK with this option enabled: https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/ThirdParty/HDF5/src/CMakeLists.txt#L27

Do you have a small block of code to reproduce the issue?

**Comment by nslay** [comment]
This is a small piece of an image batcher used for training neural networks. There's more following it that post-processes the loaded images, but that's unimportant here.

Sorry for all the extra .copy()s... I had numpy memory corruption issues too (diagnosed using MD5 sums)! Yes, you can still shoot yourself in the foot C-style in Python apparently. It just seems harder to know that you're shooting yourself in the foot because Python hides all the details.

```python
    def _load_patients(self, patientIds):
        volumeMaskPairs = []

        def load_one_patient(patientId):
            imageFiles = [ f"normalized{i+1}_aligned.nii.gz" if i > 0 else "normalized_aligned.nii.gz" for i in range(self.numChannels) ]

            if "normalized8_aligned.nii.gz" == imageFiles[-1]:
                imageFiles[-1] = "image8_aligned.nii.gz"

            imageFiles = [ os.path.join(self.dataRoot, "Images", patientId, imageFile) for imageFile in imageFiles ]


            maskFile = os.path.join(self.dataRoot, "Masks", patientId, "mask_aligned.nii.gz")

            sitkVolumes = [ LoadImage(imageFile, use_nib=True) for imageFile in imageFiles ]

            assert all(sitkVolumes), str(patientId)

            npVolumes = [ sitk.GetArrayViewFromImage(sitkVolume) for sitkVolume in sitkVolumes ]

            assert all([npVolumes[0].shape == npVolume.shape for npVolume in npVolumes])

            sitkMask = LoadMask(maskFile, numClasses=self.numClasses, dilateUnknown=self.dilateUnknown, use_nib=True)

            assert sitkMask is not None

            npMask = sitk.GetArrayViewFromImage(sitkMask).astype(np.int32) # XXX: Prevent issue with -1
            npMask[npMask == 255] = -1
            npMask[np.logical_and(npVolumes[0] < 253, npMask == 1)] = -1

            for npVolume in npVolumes:
                npMask[npVolume < -188] = -1

            npVolumes = [ npVolume[None, None, ...].copy() for npVolume in npVolumes ]
            npMask = npMask[None, ...].copy()

            return np.concatenate(tuple(npVolumes), axis=1), npMask

        with ThreadPoolExecutor(max_workers=min(8, self.batchSize)) as exe:
            volumeMaskPairs = list(exe.map(load_one_patient, patientIds))

```

Here's `LoadImage` (now modified to workaround this race using nibabel).

```python
def LoadImage(path, use_nib=False):
    try:
        if use_nib:
            import nibabel as nib
            img = nib.load(path)
            img = np.asarray(img.dataobj)
            img = img.transpose(2,1,0)
            img = sitk.GetImageFromArray(img)
            return img
        else:
            return sitk.ReadImage(path)
    except:
        return None
```

`LoadMask` calls this in addition to doing some complicated re-labeling of masks based on ITK-SNAP label text files.

**Comment by blowekamp** [comment]
Can you create a minimal [reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)?

**Comment by nslay** [comment]
Sure. But I want you to understand that this problem is not strictly reproducible.

For example, I have hundreds of experiments using 4 image channels (1 nifti file per channel) that sparsely crash due to this issue. I thought failing hardware was the cause (in fact, some machines had hardware problems). Cranking this up to 8 image channels causes a crash rate of 50%, all with "Segmentation fault". All traceable back to SimpleITK (via gdb backtrace command).

Now, you may have to wait 10 seconds or 5 days for a crash... and it may not even crash.

**Comment by nslay** [comment]
Try the attached out.
[SimpleITKCrash.zip](https://github.com/SimpleITK/SimpleITK/files/9576382/SimpleITKCrash.zip)
A screenshot is included to show the random nature of the reproducibility. I even added a "No crash!" line later to show successful terminations.

**Comment by nslay** [comment]
If I change sitk.ReadImage to nib.load, I can't seem to make it crash anymore. Indeed, I have jobs that have been running for a day now without crashing using the nibabel workaround. Although, I also some have jobs using SimpleITK that have been running for a few days now... but they can still crash. I had a 1000 epoch SimpleITK-based experiment crash at about 950 epochs... just says "Segmentation fault". And I have about half of them that completed successfully over several days using SimpleITK (and more than half of them successful when using fewer nifti images to form pytorch tensors).

**Comment by blowekamp** [comment]
@nslay For that particular issue demonstrated in crash.py I was able to use the stack trace to locale the global initialization not being concurrent thread safe. This issue appears different than the reported issue of random failure doing IO of NIFT.

The code above uses `GetArrayViewFromImage` which creates a reference to SimpleITK managed memory. It can be tricky in python and NumPy to keep track of the owner which can lead to segmentation fault. If you are frequently using `GetArrayViewFromImage` over the safer `GetArrayFromImage` then there may be incorrect uses or alias that could cause the problem.

If you have another minimal reproducible example of threading issue please create a new issue with the bug report.

**Comment by nslay** [comment]
I don't think that's the problem because memory is only shared within load_one_patient(). Everything returned by load_one_patient is a deep copy. The concatenate is obvious, but the LoadMask is also a deep copy because no one stores 32 bit integer masks. The astype(np.int32) performs a deep copy.

**Comment by blowekamp** [comment]
Have you tried the multi-process based `Pool` instead of the thread based one? With SimpleITK Images supporting pickling, the usage should be rather transparent. 

It may be that the C++ NIFTI library is no concurrent thread safe.

**Comment by nslay** [comment]
If I had to take a guess... I'd guess this is partly related to factories and their initialization. And ITK loves using lots and lots of factories everywhere!

**Comment by blowekamp** [comment]
Your report indicates that it is irregular after long execution periods. Once the factories are initialized and your code is running ( for hours ) the factories are not going to change.

**Comment by blowekamp** [comment]
The one issue is being addressed with the patch in ITK linked above.

We need a stack trace of the failure to narrow down the cause. To do that we need to reproduce the problem, with the SimpleITK symbols not stripped from the binary. Otherwise we are just guessing.

**Comment by nslay** [comment]
Or maybe valgrind or address sanitizer. A crash may occur somewhere that doesn't make any sense... caused by something else. I'll see what I can do for you.
